liquidity-rewards/contracts/UniswapPoolRewards.sol

```diff
@@ -1,5 +1,130 @@
+
+    event Deposited(address indexed user, uint256 amount);
+    event Withdrawn(address indexed user, uint256 amount);
+    event RewardClaimed(address indexed user, uint256 reward);
+
+    constructor(address _elimuTokenAddress, address _lpTokenAddress) {
+        lpToken = IERC20(_lpTokenAddress);
+        elimuToken = IERC20(_elimuTokenAddress);
+    }
+
+
+    // Returns the amount of rewards that correspond to each staked token
+    function rewardPerToken() public view returns (uint256) {
+        if (totalSupply() == 0) {
+            return rewardPerTokenStored;
+        }
+        return
+            rewardPerTokenStored.add(
+                block.timestamp
+                    .sub(lastUpdateTime)
+                    .mul(rewardRate)
+                    .mul(1e18)
+                    .div(totalSupply())
+            );
+    }
+
+    // Returns the amount that an account can claim
+    function rewardsEarned(address account) public view returns (uint256) {
+        return
+            balanceOf(account)
+                .mul(rewardPerToken().sub(userRewardPerTokenPaid[account]))
+                .div(1e18)
+                .add(rewards[account]);
+    }
+
+    // deposit visibility is public as overriding LPTokenWrapper's deposit() function
+    function deposit(uint256 amount) public override {
+        require(amount > 0, "Cannot stake 0");
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updateAccountReward(msg.sender);
+
+        super.deposit(amount);
+        emit Deposited(msg.sender, amount);
+    }
+
+    function withdraw(uint256 amount) public override{
+        require(amount > 0, "Cannot withdraw 0");
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updateAccountReward(msg.sender);
+
+        super.withdraw(amount);
+        emit Withdrawn(msg.sender, amount);
+    }
+
+    // Shortcut to be able to unstake tokens and claim rewards in one transaction
+    function withdrawAndClaim() external {
+        withdraw(balanceOf(msg.sender));
+        claimReward();
+    }
+
+    function claimReward() public {
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updateAccountReward(msg.sender);
+
+        uint256 reward = rewardsEarned(msg.sender);
+
+        require(reward > 0, "Nothing to claim");
+
+        rewards[msg.sender] = 0;
+        elimuToken.transfer(msg.sender, reward);
+        emit RewardClaimed(msg.sender, reward);
+    }
+
+    function depositReward(uint256 reward) external onlyOwner {
+        assert(reward > 0);
+        elimuToken.transferFrom(msg.sender, address(this), reward);
+        
+        uint256 totalReward = elimuToken.balanceOf(address(this));
+        _updateReward();
+
+        lastUpdateTime = block.timestamp;
+    }
+
+    function _updateReward() internal {
+        rewardPerTokenStored = rewardPerToken();
```

Comment:
Would it make sense to move `lastUpdateTime = block.timestamp;` into the `_updateReward()` function?
```suggestion
    }

    function _updateReward() internal {
        rewardPerTokenStored = rewardPerToken();
        lastUpdateTime = block.timestamp;
```